onnxruntime/core/graph/graph.cc

```diff
@@ -872,56 +872,80 @@ void Node::CreateSubgraph(const std::string& attr_name) {
-    graph_->SetGraphProtoSyncNeeded();                                       \
     AttributeProto a;                                                        \
-    a.set_name(attr_name);                                                   \
-    a.set_type(enumType);                                                    \
     *(a.mutable_##field()) = value;                                          \
-    attributes_[attr_name] = a;                                              \
-  };
+    AddAttributeHelper(*this, attr_name, enumType, std::move(a));            \
+  }
 
-#define ADD_LIST_ATTR_IMPL(type, enumType, field)            \
-  void Node::AddAttribute(const std::string& attr_name,      \
-                          const std::vector<type>& values) { \
-    graph_->SetGraphResolveNeeded();                         \
-    graph_->SetGraphProtoSyncNeeded();                       \
-    AttributeProto a;                                        \
-    a.set_name(attr_name);                                   \
-    a.set_type(enumType);                                    \
-    for (const auto& val : values) {                         \
-      *(a.mutable_##field()->Add()) = val;                   \
-    }                                                        \
-    attributes_[attr_name] = a;                              \
-  };
+#define ADD_ATTR_MOVE_IMPL(type, enumType, field)                       \
+  void Node::AddAttribute(const std::string& attr_name, type&& value) { \
+    AttributeProto a;                                                   \
+    *(a.mutable_##field()) = std::move(value);                          \
+    AddAttributeHelper(*this, attr_name, enumType, std::move(a));       \
+  }
+
+#define ADD_LIST_ATTR_IMPL(type, enumType, field)                 \
+  void Node::AddAttribute(const std::string& attr_name,           \
+                          const gsl::span<type const>& values) {  \
+    AttributeProto a;                                             \
+    auto* mutable_field = a.mutable_##field();                    \
+    for (const auto& val : values) {                              \
+      *(mutable_field->Add()) = val;                              \
+    }                                                             \
+    AddAttributeHelper(*this, attr_name, enumType, std::move(a)); \
+  }
 
 void Node::AddAttribute(const std::string& attr_name, const GraphProto& value) {
-  graph_->SetGraphResolveNeeded();
-  graph_->SetGraphProtoSyncNeeded();
   AttributeProto a;
-  a.set_name(attr_name);
-  a.set_type(AttributeProto_AttributeType::AttributeProto_AttributeType_GRAPH);
   *a.mutable_g() = value;
-  attributes_[attr_name] = a;
+  AddAttributeHelper(*this, attr_name, AttributeProto_AttributeType::AttributeProto_AttributeType_GRAPH, std::move(a));
+
+#if !defined(ORT_MINIMAL_BUILD)
+  // subgraph is created via deserialization and not here in a minimal build
+  CreateSubgraph(attr_name);
```

Comment:
does this subgraph creation need to happen for `AddAttribute(const string&, const span<const GraphProto>&)`?

Reply:
I feel we need to remove that span interface to begin with.

Reply:
that works too